prometheus-metrics/src/main/scala/org/http4s/metrics/prometheus/PrometheusMetricsNames.scala

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright 2018 http4s.org
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s.metrics.prometheus
+
+import cats.syntax.apply._
+import cats.syntax.either._
+import org.http4s.metrics.prometheus.PrometheusMetricsNames._
+
+/** Metric name may contain ASCII letters and digits, as well as underscores and colons.
+  * It must match the regex ([a-zA-Z_:][a-zA-Z0-9_:]*).
+  * Notice that only a single metric with a given name can be registered within a [[io.prometheus.client.CollectorRegistry]].
+  * Vice versa registering a metric with the same name more than once will yield an error.
+  *
+  * More details: https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels
+  */
+final class PrometheusMetricsNames private (
```

Comment:
I think this might be a very good use case for https://github.com/typelevel/literally which we already have as a dependency. Using `literally` would lift the vast majority of metric naming errors into a compile time failure rather than a runtime `IllegalArgumentException`, which would be good as many tests setups don't invoke metrics during testing.

It also might be useful to make a single newtype for a metric name, similar to what https://github.com/davenverse/epimetheus does. This would be where `literally` would come into play.

Reply:
From my experience, it's more flexible to have metrics' names (and other configurable stuff) in config. And therefore I've chosen here to check in runtime, not in compile. The main idea is that config loads at the startup of the application and if some metrics' names invalid - the application could decide to keep on startup or not.

Reply:
@danicheg using literally doesn't preclude runtime checking, it just permits compile time checking.